File: src/main/java/com/enonic/app/datakit/JwtGenerator.java
package com.enonic.app.datakit;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.KeyFactory;
import java.security.interfaces.RSAPrivateKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.time.Instant;
import java.util.Base64;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.osgi.service.component.annotations.Component;

import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;

@Component(immediate = true)
public class JwtGenerator
{
    private String cachedKeyPath;

    private RSAPrivateKey cachedKey;

    public String generateToken( final String subject, final String keyId, final String privateKeyPath,
                                 final int expirationSeconds )
    {
        try
        {
            final RSAPrivateKey privateKey = getPrivateKey( privateKeyPath );
            final Instant now = Instant.now();

            return JWT.create()
                .withKeyId( keyId )
                .withSubject( subject )
                .withIssuedAt( now )
                .withExpiresAt( now.plusSeconds( expirationSeconds ) )
                .sign( Algorithm.RSA256( null, privateKey ) );
        }
        catch ( Exception e )
        {
            throw new RuntimeException( "Failed to generate JWT: " + e.getMessage(), e );
        }
    }

    public String generateTokenFromKeyFile( final String keyFilePath, final int expirationSeconds )
    {
        try
        {
            final String content = new String( Files.readAllBytes( Paths.get( keyFilePath ) ) );
            final String kid = extractJsonValue( content, "kid" );
            final String principalKey = extractJsonValue( content, "principalKey" );
            final String pemKey = extractJsonValue( content, "privateKey" );

            final RSAPrivateKey privateKey = parsePrivateKey( keyFilePath, pemKey );
            final Instant now = Instant.now();

            return JWT.create()
                .withKeyId( kid )
                .withSubject( principalKey )
                .withIssuedAt( now )
                .withExpiresAt( now.plusSeconds( expirationSeconds ) )
                .sign( Algorithm.RSA256( null, privateKey ) );
        }
        catch ( Exception e )
        {
            throw new RuntimeException( "Failed to generate JWT from key file: " + e.getMessage(), e );
        }
    }

    private String extractJsonValue( final String json, final String key )
    {
        final Pattern pattern = Pattern.compile( "\"" + key + "\"\\s*:\\s*\"((?:[^\"\\\\]|\\\\.)*)\"" );
        final Matcher matcher = pattern.matcher( json );
        if ( matcher.find() )
        {
            return matcher.group( 1 )
                .replace( "\\n", "\n" )
                .replace( "\\\"", "\"" )
                .replace( "\\\\", "\\" );
        }
        throw new RuntimeException( "Key '" + key + "' not found in key file" );
    }

    private RSAPrivateKey getPrivateKey( final String path )
        throws Exception
    {
        if ( cachedKey != null && path.equals( cachedKeyPath ) )
        {
            return cachedKey;
        }

        final String pem = new String( Files.readAllBytes( Paths.get( path ) ) );
        return parsePrivateKey( path, pem );
    }

    private RSAPrivateKey parsePrivateKey( final String path, final String pem )
        throws Exception
    {
        if ( cachedKey != null && path.equals( cachedKeyPath ) )
        {
            return cachedKey;
        }

        final String base64 = pem
            .replace( "-----BEGIN PRIVATE KEY-----", "" )
            .replace( "-----END PRIVATE KEY-----", "" )
            .replaceAll( "\\s", "" );

        final byte[] decoded = Base64.getDecoder().decode( base64 );
        final PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec( decoded );
        final RSAPrivateKey privateKey = (RSAPrivateKey) KeyFactory.getInstance( "RSA" ).generatePrivate( spec );

        this.cachedKeyPath = path;
        this.cachedKey = privateKey;

        return privateKey;
    }
}

File: src/main/kotlin/com/enonic/app/datakit/JwtGenerator.kt
package com.enonic.app.datakit

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import org.osgi.service.component.annotations.Component
import java.nio.file.Path
import java.security.KeyFactory
import java.security.interfaces.RSAPrivateKey
import java.security.spec.PKCS8EncodedKeySpec
import java.time.Instant
import java.util.Base64
import kotlin.io.path.readText

private val JWT_JSON_STRING_REGEX = """"([^"]+)"\s*:\s*"((?:[^"\\]|\\.)*)"""".toRegex()

@Component(immediate = true)
class JwtGenerator {
    @Volatile
    private var cachedEntry: Pair<String, RSAPrivateKey>? = null

    fun generateToken(
        subject: String,
        keyId: String,
        privateKeyPath: String,
        expirationSeconds: Int,
    ): String {
        return try {
            val privateKey = getPrivateKey(privateKeyPath)
            signToken(subject, keyId, privateKey, expirationSeconds)
        } catch (e: Exception) {
            throw RuntimeException("Failed to generate JWT: ${e.message}", e)
        }
    }

    fun generateTokenFromKeyFile(keyFilePath: String, expirationSeconds: Int): String {
        return try {
            val content = Path.of(keyFilePath).readText()
            val keyId = extractJsonValue(content, "kid")
            val principalKey = extractJsonValue(content, "principalKey")
            val pemKey = extractJsonValue(content, "privateKey")

            val privateKey = parsePrivateKey(keyFilePath, pemKey)
            signToken(principalKey, keyId, privateKey, expirationSeconds)
        } catch (e: Exception) {
            throw RuntimeException("Failed to generate JWT from key file: ${e.message}", e)
        }
    }

    private fun signToken(
        subject: String,
        keyId: String,
        privateKey: RSAPrivateKey,
        expirationSeconds: Int,
    ): String {
        val now = Instant.now()

        return JWT.create()
            .withKeyId(keyId)
            .withSubject(subject)
            .withIssuedAt(now)
            .withExpiresAt(now.plusSeconds(expirationSeconds.toLong()))
            .sign(Algorithm.RSA256(null, privateKey))
    }

    private fun extractJsonValue(json: String, key: String): String {
        for (match in JWT_JSON_STRING_REGEX.findAll(json)) {
            if (match.groupValues[1] == key) {
                return match.groupValues[2].unescapeJsonString()
            }
        }

        throw RuntimeException("Key '$key' not found in key file")
    }

    private fun getPrivateKey(path: String): RSAPrivateKey {
        cachedKeyFor(path)?.let { return it }

        val pem = Path.of(path).readText()
        return parsePrivateKey(path, pem)
    }

    @Synchronized
    private fun parsePrivateKey(path: String, pem: String): RSAPrivateKey {
        cachedKeyFor(path)?.let { return it }

        val base64 = pem
            .replace("-----BEGIN PRIVATE KEY-----", "")
            .replace("-----END PRIVATE KEY-----", "")
            .filterNot(Char::isWhitespace)

        val decoded = Base64.getDecoder().decode(base64)
        val spec = PKCS8EncodedKeySpec(decoded)
        val privateKey = KeyFactory.getInstance("RSA").generatePrivate(spec) as RSAPrivateKey

        cachedEntry = path to privateKey

        return privateKey
    }

    private fun cachedKeyFor(path: String): RSAPrivateKey? {
        val (cachedPath, cachedKey) = cachedEntry ?: return null
        if (path != cachedPath) return null
        return cachedKey
    }
}

private val JSON_UNESCAPE_REGEX = """\\(.)""".toRegex()

private fun String.unescapeJsonString(): String =
    JSON_UNESCAPE_REGEX.replace(this) { match ->
        when (match.groupValues[1]) {
            "n" -> "\n"
            "\"" -> "\""
            "\\" -> "\\"
            else -> match.groupValues[1]
        }
    }
